### PR TITLE
Avoid redundant Pod restarts

### PR DESCRIPTION
**Pull request by jandubois**
Exported properties are no longer stored in secrets per pod, but in a secret per stateful set. It records the exported values. It also stores the `CONFIGGIN_VERSION_TAG`, which includes the chart version and the secrets generation, so it will change whenever there is a full restart of all pods.

Fissile will generate `CONFIGGIN_IMPORT_$ROLE` environment variables that reference the `CONFIGGIN_VERSION_TAG` in the secret of the exporting role. This prevents the importing pod from starting until the secret with the exported properties has been written.

This obviously requires that any mutual or cyclical imports between roles are resolved in the role manifest, either by using properties, or by patching some scripts to remove the bosh link and break the cycle. Otherwise there would be a deadlock as all roles in the cycle would wait forever for their imports to become ready.

There is no problem with cycles in imports from different jobs inside the same pod, as they are resolved simultaneously.

When properties are changed without a chart upgrade or secrets rotation, the pod templates of all importing stateful sets are annotated with the digest values of the exported values. The stateful set will automatically restart pods when the checksum of the pod spec has changed.

To avoid redundant restarts when the properties match the original deployment values, the annotations are not set (or removed if they already exist) when the digest matches the initial digest value.

### CODE REVIEW

**Review by andreas-kupries — APPROVED** [review]
Looks ok to me. Waiting for Mark and his checks.

File: lib/configgin.rb
Comment (mook-as) [inline lib/configgin.rb:73]: We should have a comment about this being needed to handle sidecar containers.

File: lib/configgin.rb
Comment (mook-as) [inline lib/configgin.rb:75]: And a comment on needing this because… something.

File: lib/configgin.rb
Comment (mook-as) [inline lib/configgin.rb:104]:
```suggestion
    new_tag = !secret.data.has_key?(version_tag)
```

File: lib/configgin.rb
Comment (mook-as) [inline lib/configgin.rb]: Hmm, I think my brain is broken this morning… I can't seem to understand that last sentence.

File: lib/configgin.rb
Comment (mook-as) [inline lib/configgin.rb]: No I don't think we need that

File: lib/configgin.rb
Comment (mook-as) [inline lib/configgin.rb]: Things will probably be clearer if we also keep this in a temporary variable.

File: lib/configgin.rb
Comment (mook-as) [inline lib/configgin.rb]: This will clobber any existing values (and that's intentional), right?

I think the ruby way is to spell this like:
```ruby
secret_data = {version_tag => ''} if new_tag # delete all existing secret values
```

File: lib/configgin.rb
Comment (mook-as) [inline lib/configgin.rb:122]: So if the secret already has a matching digest (or if the tag is new), we'll delete the digest from the annotation?

File: lib/kube_link_generator.rb
Comment (mook-as) [inline lib/kube_link_generator.rb]: This will blow up if the secret is missing right? (Assuming that's expected, document something in a comment?)

**Review by mook-as — CHANGES_REQUESTED** [review]
Also need to fix up the tests of course.

Mostly here to prevent accidental merges.

File: lib/configgin.rb
Comment (jandubois) [reply lib/configgin.rb:122]: Yes, that is what the comment above was trying to explain: There is no annotation for the property values exported by an install or update, so pods don't need to restart due to the template change.

File: lib/configgin.rb
Comment (mook-as) [reply lib/configgin.rb:75]: I think you mean `errands` here, but this is close enough.

File: lib/configgin.rb
Comment (mook-as) [reply lib/configgin.rb]: We discussed this offline; I just got confused by the GitHub diff view.

File: lib/configgin.rb
Comment (mook-as) [inline lib/configgin.rb:119]:
```suggestion
      secret.data["skiff-initial-digest-#{name}"] = encoded_digest if new_tag
```

File: lib/configgin.rb
Comment (mook-as) [inline lib/configgin.rb:121]:
```suggestion
      digests[name] = nil if secret.data["skiff-initial-digest-#{name}"] == encoded_digest
```

File: lib/configgin.rb
Comment (mook-as) [inline lib/configgin.rb:116]: Okay, trying to rephrase this just to see if I understand this correctly (no need to take my wording):

> In order to ensure that we restart pods when their imported properties are changed, we set annotations on the importing StatefulSet containing digests of the imported properties.

> On initial deployment, the StatefulSets do not have those annotations (due to the helm deployment); to avoid unnecessarily restarting those initial pods, we explicitly do _not_ set annotations for the initial values (when we have a new version tag).  In that case, pod startup ordering is controlled by the `CONFIGGIN_IMPORT_#{role}` environment variable references to the secret.

> In order to keep track of whether this is a new version tag, we place the version tag in the generated secret.

**Review by mook-as — APPROVED** [review]

File: lib/configgin.rb
Comment (jandubois) [reply lib/configgin.rb:119]: I normally like using statement modifiers like this, but in this case I prefer to write the conditions first, to make it visually more obvious.

File: lib/configgin.rb
Comment (jandubois) [reply lib/configgin.rb:121]: Same here

File: lib/configgin.rb
Comment (jandubois) [reply lib/configgin.rb:116]: > In order to keep track of whether this is a new version tag, we place the version tag in the generated secret.

We place the tag into the secret so fissile can create the `CONFIGGIN_IMPORT_role` variables that reference them, so kube will delay container creation until the secret exists.

And we use the version tag because during updates (or secrets rotation) the old tag already exists, but we don't want that version of the secrets, we have to wait until they are replace by the new ones.

**Comment by jandubois** [comment]
I've bumped the version and created a new tag. Let's not merge this until the tests have been updated and the commit has a proper commit message!

I'll just build a test GEM from the PR branch.

**Comment by mook-as** [comment]
I think we could use pre-release tags for this (if you don't like `update-stemcell.sh`); but if you've already published then meh.

**Comment by jandubois** [comment]
I've fixed up the tests and updated comments some more since the reviews.

There are also bug fixes, including bumping `kubeclient` to 4.3 to get support for JSON merge patches and also https://github.com/cloudfoundry-incubator/configgin/commit/7aa29f2647cf19df2ce36a829bd7f3545d38ed05.

All changes have been squashed, so the history does not include the experimental tags 0.19.2 to 0.19.6 anymore.

I've also included the trivial change in https://github.com/cloudfoundry-incubator/configgin/pull/108/commits/c8e110be357fde52de6ab8b96ab0ea6e2d692e32 to avoid having to update configgin again if we want to address CAP-828 in the future.

File: lib/configgin.rb
Comment (mook-as) [inline lib/configgin.rb:159]: I believe that's the same as `annotations.merge! digests`.  We can probably do that with `sts.spec.template.metadata.annotations` too, because `annotations` is definitely a plain hash.

**Review by andreas-kupries — APPROVED** [review]
LGTM.

File: spec/lib/configgin_spec.rb
Comment (mook-as) [inline spec/lib/configgin_spec.rb:120]: The naming's backwards here, no? (The digest we fetched is the actual digest, not the expected.)

File: spec/spec_helper.rb
Comment (mook-as) [inline spec/spec_helper.rb:47]:
```suggestion
    object.clear
```

File: spec/spec_helper.rb
Comment (mook-as) [inline spec/spec_helper.rb:71]: Nit: rubocop _should_ complain about using single quotes if there's no substitution.

**Review by mook-as — APPROVED** [review]
LGTM, tiny nits that I don't know if we should bother with.

**Pull request merged**
